File: backend/Models/RecurrenceRule.cs
public class RecurrenceRule
{
    public Guid Id { get; set; }

    // Core type of recurrence
    public RecurrenceFrequency Frequency { get; set; } // Daily, Weekly, Monthly, Yearly, Custom

    // Interval between occurrences (e.g., every 2 days, every 3 weeks, etc.)
    public int Interval { get; set; } = 1;

    // WEEKLY: which days of the week this occurs on (bitmask or list)
    public List<DayOfWeek> DaysOfWeek { get; set; } = new();

    // MONTHLY: specific day of month (e.g., 15) - null if using WeekOfMonth/DayOfWeekInMonth
    public int? DayOfMonth { get; set; }

    // MONTHLY (alternative): week of month (1 = first, 2 = second, -1 = last)
    public int? WeekOfMonth { get; set; }

    // MONTHLY (alternative): day of week in that week
    public DayOfWeek? DayOfWeekInMonth { get; set; }

    // YEARLY: specific month (1-12)
    public int? MonthOfYear { get; set; }

    // End conditions
    public RecurrenceEndType EndType { get; set; } // Never, AfterOccurrences, OnDate
    public int? OccurrenceCount { get; set; }
    public DateTime? EndDate { get; set; }

    // Metadata
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public enum RecurrenceFrequency
{
    None,
    Daily,
    Weekly,
    Monthly,
    Yearly,
    Custom
}

public enum RecurrenceEndType
{
    Never,
    AfterOccurrences,
    OnDate
}

File: frontend/src/types/RecurrenceRule.ts
import {
  DayOfWeek,
  RecurrenceEndType,
  RecurrenceFrequency,
} from '../constants/Enumeration/RecurrenceRuleEnum/RecurrenceRuleEnum';

export type TRecurrenceRule = {
  id: string;
  frequency: RecurrenceFrequency;
  interval: number;
  daysOfWeek: DayOfWeek[];
  dayOfMonth?: number | null;
  weekOfMonth?: number | null;
  dayOfWeekInMonth?: DayOfWeek | null;
  monthOfYear?: number | null;
  endType: RecurrenceEndType;
  occurrenceCount?: number | null;
  endDate?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type TAddRecurrenceRule = {
  frequency: RecurrenceFrequency;
  interval: number;
  daysOfWeek: DayOfWeek[];
  dayOfMonth?: number | null;
  weekOfMonth?: number | null;
  dayOfWeekInMonth?: DayOfWeek | null;
  monthOfYear?: number | null;
  endType: RecurrenceEndType;
  occurrenceCount?: number | null;
  endDate?: string | null;
};

export type TUpdateRecurrenceRule = {
  id: string;
  frequency: RecurrenceFrequency;
  interval: number;
  daysOfWeek: DayOfWeek[];
  dayOfMonth?: number | null;
  weekOfMonth?: number | null;
  dayOfWeekInMonth?: DayOfWeek | null;
  monthOfYear?: number | null;
  endType: RecurrenceEndType;
  occurrenceCount?: number | null;
  endDate?: string | null;
};
